python/ray/data/aggregate.py

```diff
@@ -57,49 +57,129 @@ def __init__(self):
+                res = min(a, r) if a is not None else r
+            return res
+
+        def merge(a1: AggType, a2: AggType) -> AggType:
+            if a2 is None:
+                return a1
+            if a1 is None:
+                return a2
+            return min(a1, a2)
+
         super().__init__(
             init=lambda k: None,
-            accumulate=(
-                lambda a, r: (on_fn(r) if a is None else min(a, on_fn(r)))),
-            merge=lambda a1, a2: min(a1, a2),
+            accumulate=accumulate,
+            merge=merge,
             name=(f"min({str(on)})"))
 
 
 class Max(AggregateFn):
     """Defines max aggregation."""
 
-    def __init__(self, on: Optional[AggregateOnT] = None):
+    def __init__(self,
+                 on: Optional[AggregateOnT] = None,
+                 ignore_nulls: bool = True):
         on_fn = _to_on_fn(on)
+
+        def accumulate(a: AggType, r: T):
+            r = on_fn(r)
+            if _is_null(r):
+                if ignore_nulls:
+                    res = a
+                else:
+                    raise ValueError(
+                        f"Encountered a null value {r} when calculating the "
+                        "max; please give ignore_nulls=True if you want nulls "
+                        "to be ignored.")
+            else:
+                res = max(a, r) if a is not None else r
+            return res
+
+        def merge(a1: AggType, a2: AggType) -> AggType:
+            if a2 is None:
+                return a1
+            if a1 is None:
+                return a2
+            return max(a1, a2)
+
         super().__init__(
             init=lambda k: None,
-            accumulate=(
-                lambda a, r: (on_fn(r) if a is None else max(a, on_fn(r)))),
-            merge=lambda a1, a2: max(a1, a2),
+            accumulate=accumulate,
+            merge=merge,
             name=(f"max({str(on)})"))
 
 
 class Mean(AggregateFn):
     """Defines mean aggregation."""
 
-    def __init__(self, on: Optional[AggregateOnT] = None):
+    def __init__(self,
+                 on: Optional[AggregateOnT] = None,
+                 ignore_nulls: bool = True):
         on_fn = _to_on_fn(on)
+
+        def accumulate(a: AggType, r: T):
+            r = on_fn(r)
+            if _is_null(r):
+                if ignore_nulls:
+                    res = a
+                else:
+                    raise ValueError(
+                        f"Encountered a null value {r} when calculating the "
+                        "mean; please give ignore_nulls=True if you want "
+                        "nulls to be ignored.")
+            else:
+                res = [a[0] + r, a[1] + 1]
+            return res
+
         super().__init__(
             init=lambda k: [0, 0],
-            accumulate=lambda a, r: [a[0] + on_fn(r), a[1] + 1],
+            accumulate=accumulate,
             merge=lambda a1, a2: [a1[0] + a2[0], a1[1] + a2[1]],
             finalize=lambda a: a[0] / a[1],
```

Comment:
If all the rows are null, we are going to have division by zero error here?

If we have a dataset {"col1": ["A", "B"], "col2": [None, 1]}, what the result of mean("col1", ignore_nulls=True) should be? {"col1": ["A", "B"], "mean()": [None, 1]}?

What about simple dataset [None, None], what the result of max should be? Currently if we call dataset.max() on an empty dataset, it raises `ValueError("Cannot compute max on an empty dataset")`. However if we call dataset.max() on an all-None dataset, it should do something differently?  I think the current implementation may raise the same ValueError in this case.

Reply:
>If all the rows are null, we are going to have division by zero error here?

Ah great catch! We should definitely handle this, I'll add a test for this as well.

>If we have a dataset {"col1": ["A", "B"], "col2": [None, 1]}, what the result of mean("col1", ignore_nulls=True) should be? {"col1": ["A", "B"], "mean()": [None, 1]}?

Do you mean `ds.groupby("col1").mean("col2", ignore_nulls=True)`? In that case, I think that what you gave would be the desired result.

>What about simple dataset [None, None], what the result of max should be? Currently if we call dataset.max() on an empty dataset, it raises ValueError("Cannot compute max on an empty dataset"). However if we call dataset.max() on an all-None dataset, it should do something differently? I think the current implementation may raise the same ValueError in this case.

I think that this should be return None to the user. I'm thinking that these are desirable semantics for aggregations + nulls:
- Mix of values and nulls - `ignore_nulls=True`: Ignore the nulls, return aggregation of values
- Mix of values and nulls - `ignore_nulls=False`: Return `None`
- All nulls: Return `None`
- Empty dataset: Return `None`

Reply:
Just some context on the current behavior of empty dataset handling: I actually didn't check the numpy and pandas behaviors but the native python one:
```
>>> max([])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: max() arg is an empty sequence
```

I also don't have much strong preference here. But whatever our decisions are, we should have test cases cover all the above four cases.